setup.py

```diff
@@ -45,6 +45,17 @@ def get_version():
         'scikit-image'
     ],
 
+    # Note, we probably do not want pycuda in the mainline.
```

Comment:
Note, this actually broke our CI when I left it in earlier. That is perfect example...

Reply:
I believe this is required from the original ASPIRE-python package under src/aspire/aspire. It might be better to left it as is before we clean up the code. For the APPLE-Picker module, we use cupy and set a way to switch from numpy to cupy with an option in src/aspire/config.ini. The switch interface is under  src/aspire/utils/numeric/ and the code example is src/aspire/apple/helper.py. It will be great if we can follow the similar way for Cov2D. By the way, the precompiled  pypi package for GPU finufft will be much easy to deal with by using the similar idea as the cpu version of pypi package. Ideally we don't want normal users to get finufft lib by recompiling.

Reply:
Thanks for taking a peek.  

Pycuda is clearly not directly installed by the requirements or setup.py, except here where I have just added it.  _We should not assume all users can install or run pycuda._  Yes, Pycuda is used in `src/aspire/aspire/em_classavg/*`, but importing and using it in ASPIRE code without actually having it as a dep is a bug for another day... this partly addressed that (first step). The fact that this isn't caught by the CI systems just means that code is not covered by tests. These are the sort of things I can help with.

I haven't looked at the APPLE-Picker code.  I think if you look at what this code does you might find it is already doing what you describe in the `config.ini`, by adding the `cufinufft` backend :).

I do not think there will ever be precompiled pypi package for CUDA C/C++ code.  It is not a thing that can easily exist because of basic software engineering reasons we talked about yesterday.  Not rehashing here, but we can chat about it again next time if you want.  Even the host C/C++ code is very challenging to python package correctly, and what was done for our py finufft, as I understand it, is not correct or following PyPUG; and this is reflected in that it only works for certain systems (which doesn't include my quite capable Fedora Intel i7 box, or anything in the rack of dell machines in my basement). I don't think it is good to release that to users as it is... I think that was just created for a few users in particular, works for them, and was a big help.  We should continue down that path, but it is not so simple (a different discussion really) to make portable python packages of low level code.  You will find instead a lot of python packaged ""CUDA"" codes are really using pycuda meta programming, and not actually written as CUDA C/C++ programs; this is the case for pynufft.  A different design...

Reply:
Yes. I believe the code under  `src/aspire/aspire/em_classavg/` is not merged. I don't think it can run well before we clean up. Somebody just included the whole folder over there. 

For the GPU hackathon, we should build a new branch such as `hackathon2019` and include the code of this PR over there. It will be quite messy after a few people working on it. We don't want the code be merged into the develop branch without cleaning up after the Hackathon.  Namely we should push to hackathon2019 temporarily.

Reply:
Not sure what you mean, the `em_classavg` code is definitely merged into the codebase, master and dev.

Yes, we will make such a branch. The hackathon is cute and everything, but it really doesn't have much to do with actual package development or improving the code.  I think of it more as a workshop for people less familiar with GPGPU programming. (Don't get me wrong, that is a good thing).

Most of the prep work for the hackathon _will_ eventually be mergable.  Currently cufinufft should settle just a little but more. This will hang out as a draft.

I am not clear what exactly we plan to do _at_ the hackathon, but that is not related to this PR. As far as this code goes, I am working on the `many` interface right now, and I may have a task for you, similar to your other work, if I don't get it done sooner. we can talk about it soon.

Reply:
I did not check that the `em_classavg` code can be run smoothly based on our environment. You can find an environment.yml under the subfolder.  It makes sense to merge this PI before we do GPU Hackathon.

Reply:
Okay, I want to update the repo address for cufinufft first. Give me five minutes and I'll push that.  Joakim should also look it over before merge into dev, hackathon or not, he has a good eye for things I miss.

Also if you want to make a adjoint-ness test (I think you were doing this for another code anyway correct?) that would keep the brutish test I made out of the tree. Though this isn't necessary.  One way or another we need a forward and backwards test... I'm not sure I will have time for that before the event myself.

Reply:
My implementation of checking adjoint operator has been pushed back to the orientation branch, see the link, https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/orientation/tests/test_PolarBasis2D.py#L211. Let me know if you need me work on it later.